src/private.rs

```diff
@@ -72,100 +61,118 @@ impl RedactFlags {
+        }
     }
+}
+impl ToString for RedactionTarget<'_> {
+    fn to_string(&self) -> String {
+        match self {
+            RedactionTarget::Debug { this, alternate: false } => format!("{:?}", this),
+            RedactionTarget::Debug { this, alternate: true } => format!("{:#?}", this),
+            RedactionTarget::Display(this) => this.to_string(),
+        }
+    }
+}
 
-    (|| {
-        if flags.fixed > 0 {
-            flags.redact_fixed(flags.fixed as usize, &mut redacted);
-            return;
+pub struct RedactionFormatter<'a> {
+    pub this: RedactionTarget<'a>,
+    pub flags: RedactFlags,
+    pub specialization: Option<RedactSpecialization>,
+}
+impl std::fmt::Debug for RedactionFormatter<'_> {
+    fn fmt(&self, fmt: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        #[cfg(feature = "toggle")]
+        if crate::toggle::get_redaction_behavior().is_plaintext() {
+            return self.this.passthrough(fmt);
         }
 
-        let redactable_string = to_redactable_string();
-
-        redacted.reserve(redactable_string.len());
-
-        // Specialize for Option<T>
-        if flags.is_option {
-            if redactable_string == "None" {
-                // We don't need to do any redacting
-                // https://prima.slack.com/archives/C03URH9N43U/p1661423554871499
-            } else if let Some(inner) = redactable_string
-                .strip_prefix("Some(")
-                .and_then(|inner| inner.strip_suffix(')'))
-            {
-                redacted.push_str("Some(");
-                flags.redact_partial(inner, &mut redacted);
-                redacted.push(')');
-            } else {
-                // This should never happen, but just in case...
-                flags.redact_full(&redactable_string, &mut redacted);
+        if self.flags.fixed > 0 {
+            return RedactFlags::redact_fixed(fmt, self.flags.fixed as usize, self.flags.redact_char);
+        }
+
+        let redactable_string = self.this.to_string();
+
+        #[allow(clippy::single_match)]
+        match self.specialization {
+            Some(RedactSpecialization::Option) => {
+                if redactable_string == "None" {
+                    // We don't need to do any redacting
+                    // https://prima.slack.com/archives/C03URH9N43U/p1661423554871499
+                    return fmt.write_str("None");
+                } else if let Some(inner) = redactable_string
+                    .strip_prefix("Some(")
+                    .and_then(|inner| inner.strip_suffix(')'))
+                {
+                    fmt.write_str("Some(")?;
+                    self.flags.redact_partial(fmt, inner)?;
+                    return fmt.write_char(')');
+                } else {
+                    // This should never happen, but just in case...
+                    return self.flags.redact_full(fmt, &redactable_string);
+                }
             }
-            return;
+
+            _ => {}
```

Comment:
do we not want adding a specialization to the RedactSpecialization enum and not handling it here to be a compile time error? If so we can do
```suggestion
            None => {}
```
or even get rid of the allow and do
```
        if let Some(ref specialization) = self.specialization {
            match specialization {
                RedactSpecialization::Option => {
                    if redactable_string == "None" {
                        // We don't need to do any redacting
                        // https://prima.slack.com/archives/C03URH9N43U/p1661423554871499
                        return fmt.write_str("None");
                    } else if let Some(inner) = redactable_string
                        .strip_prefix("Some(")
                        .and_then(|inner| inner.strip_suffix(')'))
                    {
                        fmt.write_str("Some(")?;
                        self.flags.redact_partial(fmt, inner)?;
                        return fmt.write_char(')');
                    } else {
                        // This should never happen, but just in case...
                        return self.flags.redact_full(fmt, &redactable_string);
                    }
                }
            }
        }
```